File: taboosearch/src/taboosearch/tsp/TspEvaluator.scala
package taboosearch.tsp

import core._

class TspEvaluator(val graph : Graph) extends Evaluator[TspSolution] {
	
	def evaluate(route : List[Int]) : Double = {
		val n = route.length;
		var value = 0.0;
		var v = route(0);
		for (i <- 1 until (n + 1)) {
			var w = route(i % route.length);
			value += graph.weights(v)(w);
			v = w;
		}
		return value;
	}
	
	def evaluate(solution : TspSolution) : Double = {
		if (solution.move == null) {
			return evaluate(solution.route);
		}
		val (i, j) = solution.move;
		val fitness = evaluate(solution.route);
		
		val v1 = solution(i);
		val v2 = solution(j);
		val t1 = solution(i - 1);
		val w1 = solution(i + 1);
		val t2 = solution(j + 1);
		val w2 = solution(j - 1);

		val weights = graph.weights;
		val delta : Double = 
			if (Math.abs((i - j) % solution.route.length) == 1)
				(- weights(t1)(v1) - weights(v2)(t2)
				 + weights(t1)(v2) + weights(v1)(t2));
			else
				(- weights(v1)(w1) - weights(w2)(v2)
			     - weights(t1)(v1) - weights(v2)(t2)
			     + weights(w2)(v1) + weights(v2)(w1)
			     + weights(v1)(t2) + weights(t1)(v2));
		return fitness + delta;
	}
}

File: src/taboosearch/tsp/TspEvaluator.java
package taboosearch.tsp;
import readers.Graph;
import taboosearch.Evaluator;

public class TspEvaluator extends Evaluator<TspSolution> {

	private double[][] weights;
	private int n;
	
	public TspEvaluator(Graph graph) {
		this.weights = graph.getWeights();
		this.n = graph.getN();
	}
	
	public double evaluate(TspSalesmanRoute route) {
		assert n == route.length();
		
		double value = 0;
		int v = route.get(0);
		
		for (int i = 1; i < n + 1; ++i) {
			int w = route.get(i % n);
			value += weights[v][w];
			v = w;
		}
		
		return value;
	}
	
	public double evaluate(TspSalesmanRoute route, double routeCost, TspSwapMove move) {
		int i = move.getI(),
			j = move.getJ();
		
		double cost = routeCost;
		int v1 = route.get(i); 
		int v2 = route.get(j);
		int t1 = route.get((i - 1) % n);
		int w1 = route.get((i + 1) % n);
		int t2 = route.get((j + 1) % n);
		int w2 = route.get((j - 1) % n);
		
		if (Math.abs(i - j) == 1) {
			cost += - weights[t1][w2] - weights[t2][w1]
			        + weights[t1][w1] + weights[t2][w2];
		} else {
			cost += - weights[v1][w1] - weights[v2][w2]
			        - weights[v1][t1] - weights[v2][t2]
			        + weights[v1][w2] + weights[v2][w1]
			        + weights[v1][t2] + weights[v2][t1];
		}
		
		return cost;
	}
	
	public double evaluate(TspSolution solution) {
		if (solution.isLazy()) {
			return evaluate(solution.getRoute(),
							solution.getRouteCost(),
							solution.getMove());
		} else {
			return evaluate(solution.getRoute());
		}
	}
	
}